pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/validator/SchemaDataValidator.java

```
@@ -0,0 +1,85 @@
+import org.apache.pulsar.broker.service.schema.exceptions.InvalidSchemaDataException;
+import org.apache.pulsar.common.schema.KeyValue;
+import org.apache.pulsar.common.schema.SchemaData;
+
+/**
+ * A validator to validate the schema data is well formed.
+ */
+public interface SchemaDataValidator {
+
+    /**
+     * Validate if the schema data is well formed.
+     *
+     * @param schemaData schema data to validate
+     * @throws InvalidSchemaDataException if the schema data is not in a valid form.
+     */
+    static void validateSchemaData(SchemaData schemaData) throws InvalidSchemaDataException {
+        switch (schemaData.getType()) {
+            case AVRO:
+            case JSON:
+            case PROTOBUF:
+                StructSchemaDataValidator.of().validate(schemaData);
+                break;
+            case STRING:
+            case BOOLEAN:
+            case INT8:
+            case INT16:
+            case INT32:
+            case INT64:
+            case FLOAT:
+            case DOUBLE:
+            case DATE:
+            case TIME:
+            case TIMESTAMP:
+                PrimitiveSchemaDataValidator.of().validate(schemaData);
+                break;
+            case NONE:
+            case BYTES:
+                // `NONE` and `BYTES` schema is not stored
+                break;
+            case AUTO:
+            case AUTO_CONSUME:
+            case AUTO_PUBLISH:
+                throw new InvalidSchemaDataException(
+                    "Schema " + schemaData.getType() + " is a client-side schema type");
+            case KEY_VALUE:
+                KeyValue<SchemaData, SchemaData> kvSchema =
+                    KeyValueSchemaCompatibilityCheck.splitKeyValueSchemaData(schemaData);
+                validateSchemaData(kvSchema.getKey());
+                validateSchemaData(kvSchema.getValue());
+                break;
+            default:
+                break;
```

is it better to throw a InvalidSchemaDataException? because use a unknown schema type.

---

good point. will change.